pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -814,13 +819,25 @@ func newAWSCloud(config io.Reader, awsServices Services) (*Cloud, error) {
 	}
 
 	selfAWSInstance, err := awsCloud.buildSelfAWSInstance()
-	if err != nil {
+	if err != nil && (cfg.Global.VPCID == "" || cfg.Global.KubernetesClusterTag == "") {
 		return nil, err
+	} else if err != nil {
+		// When the master is running on a different cloud provider or on-premise
+		// build up and dummy instance and use the vpcID from the nodes account
+		glog.Warningf("Cannot detect an AWS Instance")
```

Comment:
Can I suggest that in the `VPCID && KubenetesClusterTag` case that you skip `buildSelfAwsInstance` entirely, and just immediately build the dummy instance?  For example, it's not crazy to think you might want to run your masters in a separate account on AWS, and I think you always want a dummy instance.

Having an alternative branch for this configuration will also make this simpler I think!